crates/cargo-bsp/src/cargo_communication/cargo_types/cargo_command.rs

```diff
@@ -1,29 +1,65 @@
-use std::path::PathBuf;
+use bsp_types::BuildTargetIdentifier;
+use log::warn;
+use std::path::Path;
 use std::process::Command;
 
+use crate::project_model::target_details::TargetDetails;
 use bsp_types::requests::{CompileParams, RunParams, TestParams};
 
 pub trait CreateCommand {
     fn origin_id(&self) -> Option<String>;
 
-    fn create_command(&self, root: PathBuf) -> Command;
+    fn create_command(
+        &self,
+        root: &Path,
+        get_target_details: impl Fn(&BuildTargetIdentifier) -> Option<TargetDetails>,
+    ) -> Command;
+}
+
+fn target_ids_to_args(
+    target_id: Vec<BuildTargetIdentifier>,
+    get_target_details: impl Fn(&BuildTargetIdentifier) -> Option<TargetDetails>,
+) -> String {
+    let targets_details = target_id.iter().map(|id| {
+        get_target_details(id).unwrap_or_else(|| {
+            warn!("Target details not found for: {:?}", id);
+            TargetDetails::default()
+        })
+    });
+    targets_details
+        .map(|t| {
+            format!(
+                "--package {} --{} {} {}",
+                t.package_name.as_str(),
+                t.kind.to_string().as_str(),
+                t.name.as_str(),
+                t.get_enabled_features_str().as_str(),
+            )
+        })
+        .collect::<Vec<String>>()
+        .join(" ")
+}
+
+fn create_exec_command(exec_cmd: &str, root: &Path, targets_args: &str) -> Command {
```

Comment:
Imo lepszą nazwą na exec_cmd jest command_type. Jak zrobisz rebase, to można w sumie te stringi build, run, target dać jako stałe, bo one też będą w unit graphie wykorzystywane

Reply:
Poprawione